Add grpc request and response  metadata

samimusallam opened the pull request:
Fixes #3358 

## Short description of the changes

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

I assert span results of a call to a unary method which contains custom metadata in both the request and response.

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added
- [x] Documentation has been updated

osherv commented on experimental/packages/opentelemetry-instrumentation-grpc/src/grpc-js/index.ts:
```diff
@@ -334,4 +349,19 @@ export class GrpcJsInstrumentation extends InstrumentationBase {
+        captureResponseMetadata: metadataCapture('response', config.metadataToSpanAttributes?.server?.responseMetadata ?? [])
+      },
+    };
+  }
```
why is this transformation needed?

samimusallam replied:
to make sure no unexpected errors occur
I initialize with empty arrays if the configuration doesn't contain any metadata

osherv commented on experimental/packages/opentelemetry-instrumentation-grpc/test/helper.ts:
```diff
@@ -15,16 +15,17 @@
   context,
-  SpanKind,
-  propagation, trace,
+  // SpanKind,
```
Remove this line

samimusallam replied:
Fixed :)

osherv commented on experimental/packages/opentelemetry-instrumentation-grpc/test/helper.ts:
```diff
@@ -456,11 +495,61 @@ export const runTests = (
+
+      assertPropagation(serverSpan, clientSpan);
+
+      if (attributesValidation?.clientAttributes) {
+        for (const key in attributesValidation.clientAttributes) {
+          const expected = attributesValidation.clientAttributes[key] as string;
+          const actual = clientSpan.attributes[key] as string;
+
+          assert.equal(actual, expected);
+        }
+      }
+
+      if (attributesValidation?.serverAttributes) {
+        for (const key in attributesValidation.serverAttributes) {
+          const expected = attributesValidation.serverAttributes[key];
+          const actual = serverSpan.attributes[key];
+
+          assert.equal(actual, expected);
+        }
+      }
+    };
+
```
Consider make this function as common util function since we're going to add lot more maps to attributes :)

samimusallam replied:
Is there a common place to move it to so we can access this in other packages? or should I put it in a utils class in the grpc instrumentation only?

vmarchaud replied:
I think he meant in the grpc instrumentation as its specific to it, however i believe it can be done in another PR that add more attributes map to check.

samimusallam replied:
Sounds good! We can move it when needed

samimusallam commented:
Please let me know if I added the changes to the correct changelog as I see there's another one in opentelemetry-js/experimental :)

vmarchaud commented:
> Please let me know if I added the changes to the correct changelog as I see there's another one in opentelemetry-js/experimental :)

You  shoud move the entry to `experimental/changelog.md` since the package you modified is in `experimental`

samimusallam commented:
> > Please let me know if I added the changes to the correct changelog as I see there's another one in opentelemetry-js/experimental :)
> 
> You shoud move the entry to `experimental/changelog.md` since the package you modified is in `experimental`

Moved

vmarchaud commented on experimental/packages/opentelemetry-instrumentation-grpc/test/helper.ts:
```diff
@@ -836,5 +927,66 @@ export const runTests = (
       });
     });
+
+    describe('Test capturing metadata', () => {
```
I would prefer to have one test with grpc and grpc-js so we are sure this work with each version

vmarchaud commented:
cc @open-telemetry/javascript-approvers if someone have some time to review this :)

legendecas commented on experimental/packages/opentelemetry-instrumentation-grpc/src/utils.ts:
```diff
@@ -110,3 +110,29 @@ export const _extractMethodAndService = (name: string): { service: string, metho
+export function metadataCapture(type: 'request' | 'response', metadata: string[]) {
+  const normalizedMetadataAttributes = new Map(metadata.map(value => [value.toLowerCase(), value.toLowerCase().replace(/-/g, '_')]));
+
+  return (span: Span, getMetadata: (key: string) => undefined | string | string[] | Buffer) => {
```
I didn't find the necessity to define a getter for the grpc metadata field.

```suggestion
  return (span: Span, metadata: grpc.Metadata) => {
```

This function can get the metadata instance and retrieve the metadata pairs with `metadata.get(key)`.

samimusallam replied:
True, made things a lot more readable without the getter, Fixed.

legendecas commented on experimental/packages/opentelemetry-instrumentation-grpc/src/grpc-js/index.ts:
```diff
@@ -298,8 +306,15 @@ export class GrpcJsInstrumentation extends InstrumentationBase {
           });
+
+        instrumentation._metadataCapture.client.captureRequestMetadata(span, (metadataKey: string) => {
+          const metadataMap = metadata.getMap();
```
We should be able to retrieve the metadata value by `metadata.get(key)`.

```suggestion
          return metadata.get(metadataKey);
```

samimusallam replied:
No need for this map after getting rid of the metadata getter

legendecas reviewed (commented):
Thank you for working on this!

samimusallam commented on experimental/packages/opentelemetry-instrumentation-grpc/src/grpc-js/clientUtils.ts:
```diff
@@ -69,6 +69,7 @@ export function getMethodsToWrap(
  * span on callback or receiving an emitted event.
  */
 export function makeGrpcClientRemoteCall(
+  metadataCapture: any,
```
Added a new structure for this :)

samimusallam commented on experimental/packages/opentelemetry-instrumentation-grpc/src/utils.ts:
```diff
@@ -14,10 +14,10 @@
 import type * as grpcTypes from 'grpc';
 import type * as grpcJsTypes from '@grpc/grpc-js';
-import { IgnoreMatcher } from './types';
+import {IgnoreMatcher} from './types';
```
Done!

legendecas commented on experimental/packages/opentelemetry-instrumentation-grpc/src/grpc-js/types.ts:
```diff
@@ -59,3 +62,10 @@ export type MakeClientConstructorFunction = typeof grpcJs.makeGenericClientConst
+
+export type metadataCaptureType = {
+  client: {
+    captureRequestMetadata: (span: Span, metadata: grpcJsTypes.Metadata | grpcTypes.Metadata) => void,
```
It seems like the type is been duplicated in `experimental/packages/opentelemetry-instrumentation-grpc/src/grpc/types.ts`.

As the metadata is declared to accept both `grpcJsTypes.Metadata` and `grpcTypes.Metadata`, we can move this type declaration to the package root `experimental/packages/opentelemetry-instrumentation-grpc/src/types.ts`.

samimusallam replied:
True it is common between the two, will move it to `src/types.ts`

legendecas reviewed (commented):
LGTM % one nit
